compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -743,6 +753,117 @@ fn find_opaque_ty_constraints_for_tait(tcx: TyCtxt<'_>, def_id: LocalDefId) -> T
     hidden.ty
 }
 
+#[instrument(skip(tcx), level = "trace", ret)]
+fn may_define_opaque_type<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    opaque_def_id: LocalDefId,
+) -> bool {
+    if tcx.is_descendant_of(opaque_def_id.to_def_id(), def_id.to_def_id()) {
+        // If the opaque type is defined in the body of a function, that function
+        // may constrain the opaque type since it can't mention it in bounds.
+        return true;
+    }
+
+    if tcx.is_closure(def_id.to_def_id()) {
+        return may_define_opaque_type(tcx, tcx.local_parent(def_id), opaque_def_id);
+    }
+
+    let param_env = tcx.param_env(def_id);
+
+    trace!(parent = ?tcx.parent(opaque_def_id.to_def_id()));
+    fn has_tait<'tcx>(
+        val: impl TypeVisitable<TyCtxt<'tcx>>,
+        opaque_def_id: LocalDefId,
+        tcx: TyCtxt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+    ) -> bool {
+        struct Visitor<'tcx> {
+            opaque_def_id: DefId,
+            tcx: TyCtxt<'tcx>,
+            seen: FxHashSet<Ty<'tcx>>,
+            param_env: ty::ParamEnv<'tcx>,
+        }
+        impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for Visitor<'tcx> {
+            type BreakTy = ();
+
+            #[instrument(skip(self), level = "trace", ret)]
+            fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+                // Erase all lifetimes, they can't affect anything, but recursion
+                // may cause late bound regions to differ, because the type visitor
+                // can't erase them in `visit_binder`.
+                let t = t.fold_with(&mut ty::fold::BottomUpFolder {
+                    tcx: self.tcx,
+                    ct_op: |c| c,
+                    ty_op: |t| t,
+                    lt_op: |_| self.tcx.lifetimes.re_erased,
```

Comment:
shouldn't erase bound regions here, only escaping bound regions :thinking:

Reply:
you could implement `fn fold_binder` to call `https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.replace_bound_vars_uncached`, replacing bound vars with `self.tcx.lifetime.re_erased` and eagerly replace all free regions that way